presto-main/src/main/java/io/prestosql/sql/ParameterUtils.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql;
+
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.sql.tree.AstVisitor;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.Node;
+import io.prestosql.sql.tree.NodeRef;
+import io.prestosql.sql.tree.Parameter;
+
+import java.util.Map;
+
+public class ParameterUtils
+{
+    private ParameterUtils() {}
+
+    public static Map<NodeRef<Expression>, Integer> parameterExtractor(Node node)
+    {
+        return new AstVisitor<Map<NodeRef<Expression>, Integer>, Void>()
+        {
+            private int position;
+            @Override
+            public Map<NodeRef<Expression>, Integer> visitNode(Node node, Void context)
+            {
+                ImmutableMap.Builder<NodeRef<Expression>, Integer> builder = ImmutableMap.builder();
+                for (Node child : node.getChildren()) {
+                    builder.putAll(this.process(child));
+                }
+                return builder.build();
+            }
+
+            @Override
+            public Map<NodeRef<Expression>, Integer> visitParameter(Parameter parameter, Void context)
+            {
+                return ImmutableMap.of(NodeRef.of(parameter), position++);
```

Comment:
This will produce an ordering based on traversal order. We need to:
1) Collect all the Parameter nodes
2) Sort them based on their lexical order (`Parameter.getLocation()`)